src/dict.c

```diff
@@ -663,34 +663,37 @@ dictEntry *dictGetRandomKey(dict *d)
  * at producing N elements, and the elements are guaranteed to be non
  * repeating. */
 unsigned int dictGetRandomKeys(dict *d, dictEntry **des, unsigned int count) {
-    int j; /* internal hash table id, 0 or 1. */
+    int j = 0; /* internal hash table id, 0 or 1. */
     unsigned int stored = 0;
 
+    if (dictSize(d) == 0) return 0;
+    if (dictIsRehashing(d)) _dictRehashStep(d);
     if (dictSize(d) < count) count = dictSize(d);
-    while(stored < count) {
-        for (j = 0; j < 2; j++) {
-            /* Pick a random point inside the hash table 0 or 1. */
-            unsigned int i = random() & d->ht[j].sizemask;
-            int size = d->ht[j].size;
-
-            /* Make sure to visit every bucket by iterating 'size' times. */
-            while(size--) {
-                dictEntry *he = d->ht[j].table[i];
-                while (he) {
-                    /* Collect all the elements of the buckets found non
-                     * empty while iterating. */
-                    *des = he;
-                    des++;
-                    he = he->next;
-                    stored++;
-                    if (stored == count) return stored;
-                }
-                i = (i+1) & d->ht[j].sizemask;
-            }
-            /* If there is only one table and we iterated it all, we should
-             * already have 'count' elements. Assert this condition. */
-            assert(dictIsRehashing(d) != 0);
+    /* Pick a random point inside the hash table. */
+    int i = random() & d->ht[j].sizemask;
+    /* when rehashing, if we got an index that was already moved, go to the other hash table */
+    if (i < d->rehashidx) j = 1, i &= d->ht[j].sizemask;
+    int size = d->ht[j].size;
+    /* Make sure to visit every bucket by iterating 'size' times. */
+    while(size--) {
```

Comment:
count can't be > d->ht[0].size since the beginning of the function does:
if (dictSize(d) < count) count = dictSize(d);
anyway, i guess the comment /* Never reached. */ should be removed.